Serde json schema

selfup opened the pull request:
This is a slight feature since the last [pull request](https://github.com/pindell-matt/fe_bucket/pull/7/commits/3f1ab243d274a639c4fef08a9c956512a3e3154d) of @cite-reader got merged in.

`drop_table` function, and a slight robustness test.
- The user defined schema seems to be super easy to set up and the lib.rs takes it in with no issues.
- The next steps would be to add `append` `find_by_id` `delete_by_id` and a `where` similar to `selfup-rejs`
- These functions will only look at the records prop from the Data struct

selfup committed:
need to switch to nightly to get started

selfup committed:
for review

selfup committed:
T not Any

cite-reader committed:
Make use of serde::ser::Serialize

The essential functionality of `create_table` is implemented by calling
into `serde_json::to_string`. The bounds on this function *require* that
its argument implement `Serialize`, and so `create_table` must demand at
least that. It turns out we need no additional power, to `T: Serialize`
is the correct bound.

cite-reader committed:
Make signatures more idiomatic

These changes give us type signatures that Rustaceans are more likely to
expect.

Since all we need of the `Serializable` parameter to `create_table` is a
reference we can pass to `serde_json::to_string`, it would be rude to
take ownership of it.

Essentially all filesystem-related functions make use of `AsRef<Path>`,
which provides the caller additional flexibility. We might as well
follow their example. This also opens up the use of `Path::join`, which
is good because treating paths as strings gives me hives.

cite-reader committed:
Additional idiomatic changes

Once we use `Path::join` in `create_table`, `db_table` becomes a
`PathBuf` and `exists` is available directly.

Since what we're writing to `buffer` (actually a handle to the
underlying file) is already a string, we don't need to go through the
formatting machinery; we can blit the bytes directly.

Commit:
Merge pull request #5 from cite-reader/serde_json_schema

Improve serializer

selfup committed:
can create the db dir if not already there

selfup committed:
can read table and compare data

selfup committed:
pushing while I look for a charger

selfup committed:
works - now we need to be able to change specific data or figure out if we want to append data

selfup committed:
formatting clean up

selfup committed:
match statement formatting

selfup committed:
no more warnings

selfup committed:
0 issues with racer

selfup committed:
pushing up for review

selfup committed:
can create a table with metadata

selfup committed:
no warnings - table can be created - update table and append data next

selfup committed:
test passing - create and update table are now a thing with the new structure

selfup committed:
private pbulic interface, extracted logic into private function - refactor

selfup committed:
added dependency until code is ready and we are comfortable

cite-reader committed:
Remove unboxed_closures from feature list

This has been the default behavior since Rust 1.0, at least.

cite-reader committed:
Unsilence unused_must_use, then squash warning

A failure of `create_db_dir` indicates a serious problem; barging ahead
will inevitably cause a problem later, but any error reporting there
would inevitably point to the wrong cause. Fail fast for better
diagnostics.

cite-reader committed:
Correct error handling of directory creation

We're a library, so emitting a diagnostic message on STDOUT isn't our
decision to make.

Furthermore, the existing code behaved like it was 1996 VBScript with On
Error Resume Next. Ha ha ha no.

cite-reader committed:
Work in slices, not owned Strings, where possible

Any copies we have to make are now made exactly where we need them; this
will also make it easier to have zero-copy internals, later.

cite-reader committed:
Merge remote-tracking branch 'refs/remotes/upstream/serde_json_schema' into serde_json_schema

 Conflicts:
	src/lib.rs

cite-reader committed:
Clean up tests

Idiomatically, tests get their own module.

Also, since we're on Nightly, we can make use of the built-in
benchmarking infrastructure. `Bencher` will make decisions for us about
how many times to run the code under test, and if it's ever improved to
give us more detailed statistics we get those for free.

Commit:
Merge pull request #6 from cite-reader/serde_json_schema

Housekeeping

selfup committed:
added drop table - good solid check to make sure it works

selfup commented:
Feel free to squash if you agree with this as well.

selfup commented:
Got append working. Testing was difficult because HashMaps do not insert in order. So I made some simple assertions using `contains`.

selfup committed:
append now adds data to the records prop of the data struct and increments id's as well as next_id

selfup committed:
moved upgrade table to private fn area - made read_records private - added supporting tests

cite-reader commented on src/lib.rs:
```diff
@@ -64,8 +74,26 @@ pub fn drop_table(table: &str) -> io::Result<()> {
     Ok(())
 }
 
+#[allow(unused_must_use)]
```
This warning doesn't fire anyway.

Types marked `must_use` are generally marked as such for good reasons; I'd like to encourage you not to silence this warning.

selfup replied:
This warning absolutely fires off when using Racer

cite-reader replied:
I'll have to double check when I get off work; I could have sworn I took it
off and didn't get a warning.
On Jun 14, 2016 09:02, "Regis Boudinot" notifications@github.com wrote:

> In src/lib.rs
> https://github.com/pindell-matt/fe_bucket/pull/7#discussion_r67001196:
> 
> > @@ -64,8 +74,26 @@ pub fn drop_table(table: &str) -> io::Result<()> {
> >      Ok(())
> >  }
> > 
> > +#[allow(unused_must_use)]
> 
> This warning absolutely fires off when using Racer
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/pindell-matt/fe_bucket/pull/7/files/9fd65f447f762eaf32210c741f5d0919b1789a11..e40706def2b4330be6d641d24b18819d8fb28f26#r67001196,
> or mute the thread
> https://github.com/notifications/unsubscribe/AEAKew4aU86unJXMrPCGBEpo1BF2uRdNks5qLtCEgaJpZM4I0zw8
> .

selfup replied:
No worries! Thanks for being diligent

cite-reader replied:
I don't get `unused_must_use` warnings for this function, as written.

selfup replied:
Ok unused is gone.

selfup replied:
Here is on OSX: https://cloud.githubusercontent.com/assets/9837366/16052138/0f8b321c-3220-11e6-8d02-fd47be97472a.png

Windows: https://cloud.githubusercontent.com/assets/9837366/16051953/39e326d8-321f-11e6-8d34-94bad0589c10.png

Slightly different

selfup replied:
Here is the new diff: https://github.com/pindell-matt/fe_bucket/pull/7/commits/b6a1b369aa86020b74a750ceaad0abfff9fcd300

I still get the `unused or unknown feature, #[warn(unused_features)] on by default
src\lib.rs:1 #![feature(custom_derive, plugin, test)]` warning though

cite-reader commented on src/lib.rs:
```diff
@@ -64,8 +74,26 @@ pub fn drop_table(table: &str) -> io::Result<()> {
+
 // private functions and tests
 
+#[allow(dead_code, unused_must_use)]
```
`dead_code` doesn't fire for `pub` functions in library crates, and there isn't otherwise dead code in this function. Also, there aren't any `must_use` values in this function body that are getting dropped on the floor.

selfup replied:
Well this is a private function. I would not want a user to be able to use this if they did not know what they were doing.

selfup replied:
I'll see how I can please Racer in the mean time

cite-reader commented on src/lib.rs:
```diff
@@ -64,8 +74,26 @@ pub fn drop_table(table: &str) -> io::Result<()> {
 
+#[allow(dead_code, unused_must_use)]
+pub fn read_records<T: Serialize + Deserialize>() -> HashMap<String, T> {
+    let data: Data<T> = serde_json::from_str(&read_table("test")).unwrap();
```
Rust is smart enough to infer the type parameter; this would usually be written `let data: Data<_>`.

selfup replied:
I thought Serde was looking for `T`?

I'll try `_`

cite-reader commented on src/lib.rs:
```diff
@@ -102,6 +111,26 @@ mod tests {
+    // This is not a benchmark - it is just to make sure this can be done correctly
+    fn it_can_create_100_tables_and_drop_them_all() {
+        for n in 1..101 {
+            let table = fmt::format(format_args!("{}", n));
```
This is equivalent to `let table = format!("{}", n);`

It's vary rare to need `format_args!` unless you're writing your own macros.

selfup replied:
Yes you are correct. I will change it over to `let table = format!("{}", n);`

selfup commented:
<img width="741" alt="screen shot 2016-06-14 at 9 52 19 am" src="https://cloud.githubusercontent.com/assets/9837366/16049827/6e1d3302-3216-11e6-9508-26fa85ba21ac.png">

This is what happens when I take that warning out.

selfup committed:
changed to let table = format!("{}", n);

selfup committed:
changed Data<T>'s to Data<_>'s for serde

selfup commented:
Ok `Data<_>` totally worked for serde.

Thanks!

selfup committed:
removed std::fmt from test suite

selfup commented:
### Just for clarity.

#### This PR is still WIP

selfup committed:
valid travis file - ready for having Matt point Travis to this Repo

selfup committed:
Merge branch 'serde_json_schema' of https://github.com/pindell-matt/fe_bucket into serde_json_schema

cite-reader committed:
Create an error type

This does not currently have fe_bucket specific problems in it; I was
focused on wiring io and serde together.

The unfortunate bit of this design is that there's no distinction
between serialization errors and deserialization errors. That's
inherited from serde_json; we could offer more precise reporting, but
that means giving up try! for serde operations.

ref #10

selfup commented:
@cite-reader 

So on my 3 year old laptop running Ubuntu 14.04 LTS I get: `65,796 ns/iter (+/- 2,925,742)`

Theory stands!

selfup committed:
removed unused allow clause kept dead_code - still get a warning about the test feature

selfup commented:
@cite-reader 

Ok all but one warning left and it seems to be: `unused or unknown feature, #[warn(unused_features)] on by default
src\lib.rs:1 #![feature(custom_derive, plugin, test)]`

Otherwise this branch should be good to go!

selfup commented:
Other than the lone warning, I deem this ready

cite-reader committed:
Squash "unused feature" warning in non-test builds

cite-reader commented:
LGTM

cite-reader committed:
Add error for nonexistent table and excise panics

The only library-specific error case I noticed in the current code is
trying to operate on a table that isn't there. That error variant
carries the name of the table around as an owned String; we could use a
slice there, but then Error would need a lifetime parameter.

Fallible operations return our Result, instead of possibly panicking.
The most invasive change to do that was read_table, which now reads the
file directly into a string rather than reading line-by-line into an
intermediate buffer and joining at the end, since I didn't want to try
to remember how Rust pronounces "sequence". This version saves us some
copies anyway.

One side-effect of the depanicking is that we don't need
allow(unused_must_use) anymore: all must_use intermediates are returned
to the caller. Except for in tests, which got _more_ panicky, but let's
be honest if any of those steps failed we wanted the test to fail anyway
right?

selfup commented:
@cite-reader Do you want to wait until your error branch gets merged?

cite-reader commented:
I'm going to be busy moving into a new apartment for the next few days and
won't have time to contribute for a while. Please handle the situation
however you think is best.
On Jun 15, 2016 23:40, "Regis Boudinot" <notifications@github.com> wrote:

> @cite-reader <https://github.com/cite-reader> Do you want to wait until
> your error branch gets merged?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pindell-matt/fe_bucket/pull/7#issuecomment-226400849>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AEAKe4kP8oDGBImuyGuhPVvn9vCUZf_2ks5qMO_ygaJpZM4I0zw8>
> .
>

selfup commented:
@cite-reader: No worries!

Enjoy your new place.

I'll see what I can do. Maybe I'll take your code and put it in this branch. Up to you

I'll talk to @pindell-matt about it.

The other alternative is that this branch can get Merged, and you can patch the errors branch so that it can be ready when you have time!

selfup committed:
Adding cite-reader lib.rs changes to see if errors branch will merge

selfup committed:
adding errors module from @cite-reader to fix conflicts

selfup committed:
removed test from feature clause

selfup committed:
Had to add test back to feature for tests to run

selfup committed:
Merge branch 'serde_json_schema' of https://github.com/pindell-matt/fe_bucket into serde_json_schema

selfup committed:
Merge pull request #11 from cite-reader/errors

Error types

selfup commented:
@cite-reader 
@pindell-matt 

I made some guesses locally on my end and pushed up until there were no more conflicts from the errors branch.

Worked like a charm!

Now this PR is ready to go!

selfup committed:
Removed uneeded dependency

selfup committed:
All functions use a taken param - read_records removed since it was not being used - clean up before merge

selfup committed:
staying under 100 chars
